src/core/globalstate.ts

```diff
@@ -45,6 +45,12 @@ export class MobXGlobals {
 	isRunningReactions = false;
 
 	/**
+	 * Current reaction (if any is running).
+	 * Reactions are run after derivations using a trampoline.
+	 */
+	currentReaction: Reaction = undefined;
```

Comment:
Good point. Same holds for `when`, but it's effect disposes the reaction anyway. Approach 1) doesn't gonna work nice, effect is not a tracked function at all, so that might only be confusing. For 2) the same, you would get a fake kind of `currentReaction`, because effect doesn't really running reaction. 

That leaves only 3. Or either of the following: pass in a second parameter to `effect`; the reaction disposer, or have a  separate state var specific for this purpose, something like `currentReactionDisposer` or something. That is a bit iffy though, as this would need to be a stack as autorun's could be nested.

Would a viable alternative to this problem be always passing in the reaction object to the callback functions of reactions? I think meteor does something similar, and then you can always write things like:

```javascript
reaction(
  (r) => {
    r.dispose()
  },
  (data,r) => {
    r.dispose()
  }
)
```

Reactions are nowhere really visible to the outside world, so it could also be that the last argument is just the disposer func. But passing the complete reaction as object seems more flexible towards the future.